### ISSUE TITLE
ZC158: PHP Warning: Use of undefined constant TOPMOST_CATEGORY_PARENT_ID with CEON URImapping

### ISSUE THREAD

**Issue by torvista** [zencart/zencart#4939]
Firstly apologies for this not being easily reproduceable.  It is a lot of work to attempt to replicate this on a vanilla site, with stock plugin files as I use the (commercial) UMM version, on 157/158 etc.
It is intended as a heads-up for developers who may discover this post-release.
Using Ceon Uri Mapping, I have long gotten debugs for TOPMOST_CATEGORY_PARENT_ID not being defined, when accessing the categories.

> #1  zen_get_parent_categories() called at [SHOPFRONT/includes/classes/class.CeonURIMappingHandler.php:363]
> #2  CeonURIMappingHandler->_handleStaticURI() called at [SHOPFRONT/includes/classes/class.CeonURIMappingHandlerBase.php:455]
> #3  CeonURIMappingHandlerBase->_handleURI() called at [SHOPFRONT/includes/classes/class.CeonURIMappingHandlerBase.php:149]
> #4  CeonURIMappingHandlerBase->__construct() called at [SHOPFRONT/includes/classes/class.CeonURIMappingHandler.php:72]
> #5  CeonURIMappingHandler->__construct() called at [SHOPFRONT/includes/autoload_func.php:47]
> #6  require(SHOPFRONT/includes/autoload_func.php) called at [SHOPFRONT/includes/application_top.php:252]
> #7  require(SHOPFRONT/includes/application_top.php) called at [SHOPFRONT/index.php:25]
> --> PHP Warning: Use of undefined constant TOPMOST_CATEGORY_PARENT_ID - assumed 'TOPMOST_CATEGORY_PARENT_ID' (this will throw an Error in a future version of PHP) in \includes\functions\functions_categories.php on line 186.

The Ceon shopfront class class.CeonURIMappingHandler.php is loaded at point 0 and instantiated at point 99 (95 in my shop for some undocumented reason).
this calls function zen_get_parent_categories which references TOPMOST_CATEGORY_PARENT_ID.
TOPMOST_CATEGORY_PARENT_ID is defined in init_category_path which is not loaded until point 160, hence the error.
While some rejigging of load points could maybe cure this (not easily, in my attempts) I am not sure if the locations are optimum for the definition of TOPMOST_CATEGORY_PARENT_ID as it is defined in both admin and shopfront separately.
I moved the shopfront one to an extra_configures file as a temporary fix but perhaps it should be defined in one place/database and so loaded earlier anyway?

**Comment by torvista** [comment]
> This is not a bug in Zen Cart.
Apparently not. But it is not a "bug" in the Ceon code either. Perhaps it highlights something that requires some attention.
The people that will come across this issue and can offer valid input inhabit this space.
Posting about unreleased software in the forum is not appropriate.

**Comment by lat9** [comment]
@torvista, thanks for the heads-up.  I've not seen such an issue on sites using the free version of Ceon URI Mapping, but will be on the look-out.

**Comment by neekfenwick** [comment]
For reference, this was mentioned on the Support thread at https://www.zen-cart.com/showthread.php?225478-Ceon-URI-Mapping-V5-0&p=1390877#post1390877 (thank you @torvista for the connection).

**Comment by mc12345678** [comment]
> > This is not a bug in Zen Cart.
> 
> Apparently not. But it is not a "bug" in the Ceon code either. Perhaps it highlights something that requires some attention. The people that will come across this issue and can offer valid input, inhabit this space. Posting about unreleased software in the forum is not appropriate.

You mean that perhaps the file should be [PSR-1](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-1-basic-coding-standard.md#1-overview) compliant by not violating the third bullet: 

- Files SHOULD either declare symbols (classes, functions, constants, etc.) or cause side-effects (e.g. generate output, change .ini settings, etc.) but SHOULD NOT do both.

Which is perhaps better explained in [Section 2.3](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-1-basic-coding-standard.md#23-side-effects). 

True though, non-compliance with the coding standard is not a bug.

**Comment by torvista** [comment]
I'm afraid I'm none the wiser.:confused:

**Comment by mc12345678** [comment]
Basically, the file currently defines a variable *and* preforms calculations both in the same file. Further the define is used in support of the local operation instead of at our before it possibly could be used. E.g. the functions, at least one of which needs that define are loaded at trigger point 60, but the/a define in support of them isn't loaded until 160... 

Add to this that the (ignored) duplicate warnings have the possibility of becoming errors in the future, preemptively working around the Undefined variable will likely lead to those future errors needing addressing when approaching that bridge. 

I'm also a little surprised that the constant is outright defined instead of defined only if not yet defined. From the PR notes it is a value that is intended to possibly modified, but as incorporated that requires modification of a core file. That core file would get overwritten in a "blind" upgrade.

**Comment by neekfenwick** [comment]
@scottcwilson a repeat of @mc12345678 point above, to try to contend your position that this isn't a bug, while also solving the PSR-1 debate, consider the order of init breakpoints:

breakpoint 60: init_general_funcs.php (from config.core.php)
  - loads functions_categories which defines functions that rely on TOPMOST_CATEGORY_PARENT_ID e.g. zen_get_parent_categories()

breakpoint 160: init_category_path.php
- defines TOPMOST_CATEGORY_PARENT_ID
- uses TOPMOST_CATEGORY_PARENT_ID when cPath is null

This ordering shows me that any code after breakpoint 60 but before 160 may call `zen_get_parent_categories()` and other fns in a buggy fashion because `TOPMOST_CATEGORY_PARENT_ID` will not be defined, hence this is a bug in zencart core code.

One logical solution to this would be to move the definition of `TOPMOST_CATEGORY_PARENT_ID` to either breakpoint 60 or before, and thus out of init_category_path.php.

From my experience it seems sensible to move this constant definition into the `configuration` table and thus have it defined in `init_db_config_read` at breakpoint 40 without needing an explicit `define()` statement in code.  To put it in `configuration` would presumably require a database migration step in zc_install, and we'd have to decide where to put it in the admin Configuration UI.  At least, I agree with @mc12345678 that it's far better to have this value not hardcoded, especially when it relates to db data (categories.categories_id).

If I put a PR together that removes the `define` from init_category_path.php and amends zc_install/sql/updates/mysql_upgrade_zencart_158.sql to insert a new `configuration` table entry, is it likely to be accepted? (I'm very new to zencart github workflow)

**Comment by torvista** [comment]
I would add that I have an autoemailer running on a nightly cron job, which calls admin application top and often/not always gets an error for missing this constant, which is odd.

**Comment by scottcwilson** [comment]
@lat9 I'm thinking we should migrate `TOPMOST_CATEGORY_PARENT_ID` out of `includes/init_includes/init_category_path.php` and `admin/includes/init_includes/init_category_path.php` into 
`includes/init_includes/init_non_db_settings.php`

Thoughts?

**Comment by celextel** [comment]
We are also having this issue with Ultimate SEO URLs enabled:
[https://www.zen-cart.com/showthread.php?229110-Undefined-constant-TOPMOST_CATEGORY_PARENT_ID&p=1391207#post1391207](https://www.zen-cart.com/showthread.php?229110-Undefined-constant-TOPMOST_CATEGORY_PARENT_ID&p=1391207#post1391207)

**Comment by proseLA** [comment]
> 
> If I put a PR together that removes the `define` from init_category_path.php and amends zc_install/sql/updates/mysql_upgrade_zencart_158.sql to insert a new `configuration` table entry, is it likely to be accepted? (I'm very new to zencart github workflow)

in answer to your question, who knows.

i'm guessing there will be some letter designate releases for v158.  in the past, these changes did NOT involve any new database updates.  so while your PR would/could get accepted, if the past is an indication of the future, it would not see the light of day until the next numerical release.

given the severity of the bug and its effect on the community, i think that is why scott is proposing adding it to:
```
includes/init_includes/init_non_db_settings.php
```

**Comment by scottcwilson** [comment]
@proseLA The message you are replying to appears to have been deleted, so I'm not sure who suggested it, but your response is correct - a database modification is not the way to go here. 

I'll go ahead and build the PR as described above in the next day or two but please go ahead and modify your own cart using this technique if you're in a hurry.

**Comment by proseLA** [comment]
@scottcwilson i only quoted part of the message.  it has not been deleted.  it is at the end of the comment here:

https://github.com/zencart/zencart/issues/4939#issuecomment-1301890883

from @neekfenwick

**Comment by scottcwilson** [comment]
PR created. https://github.com/zencart/zencart/pull/5371 

init_non_db_settings.php is run at breakpoint 45 so it should be early enough.

**Comment by neekfenwick** [comment]
> given the severity of the bug and its effect on the community, i think that is why scott is proposing adding it to:
> 
> ```
> includes/init_includes/init_non_db_settings.php
> ```

I see. I haven't encountered init_non_db_settings.php before, looks like a very curious file, but yes that would solve the problem :) (at least until someone points out the PSR-1 violation :P )

**Comment by scottcwilson** [comment]
> I haven't encountered init_non_db_settings.php before, looks like a very curious file ...

New to 1.5.8! We made an effort to migrate defines which are not language related out of the language files.

**Comment by neekfenwick** [comment]
> New to 1.5.8! We made an effort to migrate defines which are not language related out of the language files.

In the past I've used a file called something like `mysite_datafiles.php`, probably in `includes/extra_datafiles` I forget offhand, which contains a fat load of `define()` statements that sets up constants specific to our site.  Doesn't that system (of arbitrary number of php files according to your site) do what init_non_db_settings.php is trying to do?

**Comment by scottcwilson** [comment]
Sounds like it.

**Comment by scottcwilson** [comment]
Saw this issue on a non-Ceon site.  Updating known 158 bugs list. 

Log:
```
[12-Dec-2022 02:24:56 UTC] PHP Fatal error:  Uncaught Error: Undefined constant "TOPMOST_CATEGORY_PARENT_ID" in /home/client/public_html/site/_livesite/includes/functions/functions_categories.php:186
Stack trace:
#0 /home/client/public_html/site/_livesite/includes/functions/functions_categories.php(639): zen_get_parent_categories(Array, '1')
#1 /home/client/public_html/site/_livesite/includes/classes/shopping_cart.php(1913): zen_get_generated_category_path_rev('1')
#2 /home/client/public_html/site/_livesite/includes/main_cart_actions.php(47): shoppingCart->actionAddProduct('shopping_cart', Array)
#3 /home/client/public_html/site/_livesite/includes/init_includes/init_cart_handler.php(42): require('/home/client...')
#4 /home/client/public_html/site/_livesite/includes/autoload_func.php(40): require_once('/home/client...')
#5 /home/client/public_html/site/_livesite/includes/application_top.php(237): require('/home/client...')
#6 /home/client/public_html/site/_livesite/index.php(25): require('/home/client...')
#7 {main}
  thrown in /home/client/public_html/site/_livesite/includes/functions/functions_categories.php on line 186

[12-Dec-2022 02:24:56 UTC] Request URI: /index.php?main_page=product_info&products_id=13%3A83a5861e6cc1796e373a10f761cf815c&number_of_uploads=0&action=add_product, IP address: 98.213.30.194

```

**Comment by neekfenwick** [comment]
> #3 /home/client/public_html/site/_livesite/includes/init_includes/init_cart_handler.php(42): require('/home/client...')

`init_cart_handler` is at breakpoint 140, well after your change in `init_non_db_settings` at breakpoint 40.  Wouldn't your already-merged PR #5371 fix this?

**Comment by scottcwilson** [comment]
The intention of my comment was to note that this PR may be required for non-Ceon sites as a patch to the originally released 1.5.8.

**Comment by neekfenwick** [comment]
Ah :) Yes, indeed.  Thanks.
